File: backtracking/17.letter-combinations-of-a-phone-number.py
from typing import List


# @leet start
class Solution:
    def __init__(self) -> None:
        self.res = []
        self.mapping = "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"

    def dfs(self, digits: str, path: list[str]):
        if len(path) == len(digits):
            self.res.append("".join(path))
            return
        for c in self.mapping[int(digits[len(path)])]:
            path.append(c)
            self.dfs(digits, path)
            path.pop()

    def letterCombinations(self, digits: str) -> List[str]:
        if digits == "":
            return []
        self.dfs(digits, [])
        return self.res


# @leet end

File: Backtracking/17.letter-combinations-of-a-phone-number.go
package leetcode

// @leet start
var (
	res      []string
	path     []rune
	keyboard = map[rune]string{
		'2': "abc",
		'3': "def",
		'4': "ghi",
		'5': "jkl",
		'6': "mno",
		'7': "pqrs",
		'8': "tuv",
		'9': "wxyz",
	}
)

func letterDfs(digits string) {
	if len(path) == len(digits) {
		res = append(res, string(path))
		return
	}
	curRune := rune(digits[len(path)])
	for _, r := range keyboard[curRune] {
		path = append(path, r)
		letterDfs(digits)
		path = path[:len(path)-1]
	}
}

func letterCombinations(digits string) []string {
	res = []string{}
	if digits == "" {
		return res
	}
	letterDfs(digits)
	return res
}
